### PR TITLE
BB-119: add lifecycle metrics

### PR DESCRIPTION
**Pull request by alexanderchan-scality**

### CODE REVIEW

**Review by miniscruff — APPROVED** [review]

File: extensions/lifecycle/LifecycleMetrics.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+const { ZenkoMetrics } = require('arsenal').metrics;
+
+const LIFECYCLE_LABEL_ORIGIN =  'origin';
+const LIFECYCLE_LABEL_PROCESS = 'process';
+const LIFECYCLE_LABEL_OP = 'op';
+const LIFECYCLE_LABEL_STATUS = 'status';
+
+const conductorLastBatchTime = ZenkoMetrics.createGauge({
+    name: 'lifecycle_conductor_last_batch_time',
+    help: 'Timestamp of latest lifecycle batch',
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/LifecycleMetrics.js]: Since it marks the (end of the) start of the batch, it  should be in the metric name and description to avoid confusion (like if the last timestamp was 1h ago, someone could expect to see some results when in fact it could still be running).

File: extensions/lifecycle/LifecycleMetrics.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    labelNames: [
+        LIFECYCLE_LABEL_ORIGIN
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/LifecycleMetrics.js]: Is the origin label standard in artesca / already used by other artesca components? As usual, trying to minimize cardinality in prometheus, this would benefit from merging with process, using `origin=lifecycle-expiration` instead of `origin=lifecycle,process=expiration`.

Also the origin itself is in the counter name so no need for an extra label.

File: extensions/lifecycle/LifecycleMetrics.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        LIFECYCLE_LABEL_PROCESS,
+    ],
+});
+
+const lifecycleKafkaPublish = ZenkoMetrics.createCounter({
+    name: 'lifecycle_kafka_publish',
+    help: 'Total number messages published by lifecycle processes',
+    labelNames: [
+        LIFECYCLE_LABEL_ORIGIN,
+        LIFECYCLE_LABEL_STATUS,
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/LifecycleMetrics.js]: Since status here is binary (success vs error), it could be 2 counters, as they are easy to sum, without the label tax.

File: extensions/lifecycle/LifecycleMetrics.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    if (statusCode >= 200) {
+        return '2xx';
+    }
+
+    return '1xx';
+}
+
+class LifecycleMetrics {
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/LifecycleMetrics.js:80]: It would be wise to wrap all of these in try blocks, to avoid crashing the process in case anything (bug or runtime issue) throws, as this is unrelated to data flow.

**Review by rachedbenmustapha — DISMISSED** [review]

File: extensions/lifecycle/LifecycleMetrics.js (hunk 3 above)
Comment (alexanderchan-scality) [reply extensions/lifecycle/LifecycleMetrics.js]: ok. I can split the counter into separate counts for the different http code ranges to reduce overhead from label processing

File: extensions/lifecycle/LifecycleMetrics.js (hunk 2 above)
Comment (alexanderchan-scality) [reply extensions/lifecycle/LifecycleMetrics.js]: will remove the redundancy, if label consistency is not required between the extensions, `origin` and `op` will be the only required labels

File: extensions/lifecycle/LifecycleMetrics.js (hunk 3 above)
Comment (rachedbenmustapha) [reply extensions/lifecycle/LifecycleMetrics.js]: Here specifically for Kafka messages and only 2 outcomes possible it can be 2 counters. For HTTP status code ranges, a label is common practice and a reasonable use of cardinality.

File: extensions/lifecycle/LifecycleMetrics.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
+const { ZenkoMetrics } = require('arsenal').metrics;
+
+const LIFECYCLE_LABEL_ORIGIN =  'origin';
+const LIFECYCLE_LABEL_OP = 'op';
+const LIFECYCLE_LABEL_STATUS = 'status';
+
+const conductorLatestBatchTime = ZenkoMetrics.createGauge({
```
Comment (francoisferrand) [inline extensions/lifecycle/LifecycleMetrics.js]: time (usually) increases, is this true also for this timestamp?

If that is the case, we may use counter instead : so that prometheus gets the "intention" that this field increase monotonically, and handles reset appropriately.

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 6]
```diff
@@ -69,7 +70,24 @@ class LifecycleTask extends BackbeatTask {
...
+                    'error',
+                );
+            } else {
+                LifecycleMetrics.onKafkaPublish(
+                    null,
+                    'sendBucketEntry',
+                    'bucket',
+                    'success',
+                );
+            }
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js]: May be better to simplify code :

```suggestion
            LifecycleMetrics.onKafkaPublish(
                null,
                'sendBucketEntry',
                'bucket',
                err ? 'error' : 'success',
            );
```

Or even to pass the `err` directly to onKafkaPublish() function : and let that function handle the "mapping" to whatever metrics is needed (thus keeping the `success` and `error` private implementation details of LifecycleMetrics kcass)

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 7]
```diff
@@ -109,7 +127,24 @@ class LifecycleTask extends BackbeatTask {
      */
     _sendObjectAction(entry, cb) {
         const entries = [{ message: entry.toKafkaMessage() }];
-        this.producer.sendToTopic(this.objectTasksTopic, entries, cb);
+        this.producer.sendToTopic(this.objectTasksTopic, entries,  err => {
+            if (err) {
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js]: same

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 8]
```diff
@@ -134,13 +169,22 @@ class LifecycleTask extends BackbeatTask {
         async.waterfall([
             next => req.send((err, data) => {
                 if (err) {
+                    LifecycleMetrics.onS3Request(
+                        log,
+                        'ListObjects',
+                        'bucket',
+                        err.statusCode
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js]: similarly, may be better to always pass err to `onS3Request` : and let the implementation decide to use statusCode field or otherwise.

File: extensions/lifecycle/LifecycleMetrics.js (hunk 5 above)
Comment (alexanderchan-scality) [reply extensions/lifecycle/LifecycleMetrics.js]: the timestamp is a  unix timestamp and will be a monotonically increasing number.

**Review by naren-rajendran — APPROVED** [review]

**Comment by alexanderchan-scality** [comment]
ping

File: extensions/lifecycle/LifecycleMetrics.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
+const { ZenkoMetrics } = require('arsenal').metrics;
+
+const LIFECYCLE_LABEL_ORIGIN =  'origin';
+const LIFECYCLE_LABEL_OP = 'op';
+const LIFECYCLE_LABEL_STATUS = 'status';
+
+const conductorLatestBatchStartTime = ZenkoMetrics.createCounter({
+    name: 'lifecycle_latest_batch_start_time',
+    help: 'Timestamp of latest lifecycle batch start time',
+    labelNames: [LIFECYCLE_LABEL_ORIGIN],
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/LifecycleMetrics.js]: There can be only 1 origin for this counter.

File: extensions/lifecycle/conductor/LifecycleConductor.js [hunk 10]
```diff
@@ -21,6 +21,8 @@ const VaultClientCache = require('../../../lib/clients/VaultClientCache');
 const CredentialsManager = require('../../../lib/credentials/CredentialsManager');
 const safeJsonParse = require('../util/safeJsonParse');
 
+const { LifecycleMetrics } = require('../LifecycleMetrics');
```
Comment (rachedbenmustapha) [inline extensions/lifecycle/conductor/LifecycleConductor.js:24]: There could be a few ops captured here as well, such as list buckets, get account ids, ...

**Review by rachedbenmustapha — CHANGES_REQUESTED** [review]

File: extensions/lifecycle/LifecycleMetrics.js (hunk 5 above)
Comment (alexanderchan-scality) [reply extensions/lifecycle/LifecycleMetrics.js]: this however will need to be a gauge in order to use the `set` method

**Review by rachedbenmustapha — APPROVED** [review]
